airbyte-workers/src/main/java/io/airbyte/workers/internal/StateMetricsTracker.java

```diff
@@ -0,0 +1,241 @@
+      final List<byte[]> stateMessagesForStream = streamStateHashesAndTimestamps.get(streamNameAndNamespace);
+      startingTime = findStartingTimeStampAndRemoveOlderEntries(stateMessagesForStream, stateHash);
+    } else {
+      startingTime = findStartingTimeStampAndRemoveOlderEntries(stateHashesAndTimestamps, stateHash);
+    }
+    updateMaxAndMeanSeconds(startingTime, timeCommitted);
+  }
+
+  void addStateMessageToStreamToStateHashTimestampTracker(final AirbyteStateMessage stateMessage,
+                                                          final int stateHash,
+                                                          final Long epochTimeEmitted) {
+
+    final StreamDescriptor streamDescriptor = stateMessage.getStream().getStreamDescriptor();
+    final String streamNameAndNamespace = streamDescriptor.getName() + streamDescriptor.getNamespace();
+    final byte[] stateHashAndTimestamp = populateStateTimestampByteArray(stateHash, epochTimeEmitted);
+
+    if (streamStateHashesAndTimestamps.get(streamNameAndNamespace) == null) {
+      final List stateHashesAndTimestamps = new ArrayList<>();
+      stateHashesAndTimestamps.add(stateHashAndTimestamp);
+      streamStateHashesAndTimestamps.put(streamNameAndNamespace, stateHashesAndTimestamps);
+    } else {
+      final List<byte[]> streamDescriptorValue = streamStateHashesAndTimestamps.get(streamNameAndNamespace);
+      streamDescriptorValue.add(stateHashAndTimestamp);
+    }
+  }
+
+  void updateMaxAndMeanSeconds(final LocalDateTime startingTime, final LocalDateTime timeCommitted) {
+    final Long secondsUntilCommit = calculateSecondsBetweenStateEmittedAndCommitted(startingTime, timeCommitted);
+    if (maxSecondsBetweenStateMessageEmittedandCommitted < secondsUntilCommit) {
+      maxSecondsBetweenStateMessageEmittedandCommitted = secondsUntilCommit;
+    }
+
+    if (totalDestinationEmittedStateMessages.get() == 1) {
+      meanSecondsBetweenStateMessageEmittedandCommitted = secondsUntilCommit;
+    } else {
+      final Long newMeanSeconds =
+          calculateMean(meanSecondsBetweenStateMessageEmittedandCommitted, totalDestinationEmittedStateMessages.get(), secondsUntilCommit);
+      meanSecondsBetweenStateMessageEmittedandCommitted = newMeanSeconds;
+    }
+  }
+
+  private LocalDateTime findStartingTimeStampAndRemoveOlderEntries(final List<byte[]> stateList, final int stateHash) {
+    // iterate through each [state_hash, timestamp] in the list
+    // update the first timestamp to equal min_timestamp
+    // and remove all items from the list as we iterate through
+    // break once we reach the state hash equal to the input(destination) state hash
+    Long minTime = null;
+    final Iterator<byte[]> iterator = stateList.iterator();
+    while (iterator.hasNext()) {
+      final byte[] stateMessageTime = iterator.next();
+      final ByteBuffer current = ByteBuffer.wrap(stateMessageTime);
+      remainingCapacity += current.capacity();
```

Comment:
Why do we increase the capacity by the capacity of the buffer? I suspect it should be `BYTE_ARRAY_SIZE`.